internal/clients/kibana/slo.go

```diff
@@ -0,0 +1,171 @@
+		Description:     &s.Description,
+		Indicator:       &indicator,
+		TimeWindow:      &s.TimeWindow,
+		BudgetingMethod: (*slo.BudgetingMethod)(&s.BudgetingMethod),
+		Objective:       &s.Objective,
+		Settings:        s.Settings,
+	}
+
+	req := client.UpdateSloOp(ctxWithAuth, s.SpaceID, s.ID).KbnXsrf("true").UpdateSloRequest(reqModel)
+	slo, res, err := req.Execute()
+	if err != nil && res == nil {
+		return nil, diag.FromErr(err)
+	}
+
+	defer res.Body.Close()
+	if diags := utils.CheckHttpError(res, "Unable to update slo with id "+s.ID); diags.HasError() {
+		return nil, diags
+	}
+
+	return sloResponseToModel(s.SpaceID, slo), diag.Diagnostics{}
+}
+
+func CreateSlo(ctx context.Context, apiClient *clients.ApiClient, s models.Slo) (*models.Slo, diag.Diagnostics) {
+	client, err := apiClient.GetSloClient()
+	if err != nil {
+		return nil, diag.FromErr(err)
+	}
+
+	ctxWithAuth := apiClient.SetGeneratedClientAuthContext(ctx)
+	indicator, err := responseIndicatorToCreateSloRequestIndicator(s.Indicator)
+	if err != nil {
+		return nil, diag.FromErr(err)
+	}
+	reqModel := slo.CreateSloRequest{
+		Name:            s.Name,
+		Description:     s.Description,
+		Indicator:       indicator,
+		TimeWindow:      s.TimeWindow,
+		BudgetingMethod: slo.BudgetingMethod(s.BudgetingMethod),
+		Objective:       s.Objective,
+		Settings:        s.Settings,
+	}
+	req := client.CreateSloOp(ctxWithAuth, s.SpaceID).KbnXsrf("true").CreateSloRequest(reqModel)
+	sloRes, res, err := req.Execute()
+	if err != nil && res == nil {
+		return nil, diag.FromErr(err)
+	}
+	defer res.Body.Close()
+
+	if diags := utils.CheckHttpError(res, "Unable to create slo"); diags.HasError() {
+		return nil, diags
+	}
+
+	s.ID = sloRes.Id
+
+	return &s, diag.Diagnostics{}
+}
+
+func responseIndicatorToCreateSloRequestIndicator(s slo.SloResponseIndicator) (slo.CreateSloRequestIndicator, error) {
+	var ret slo.CreateSloRequestIndicator
+
+	ind := s.GetActualInstance()
+	switch ind.(type) { //nolint:all
+	case *slo.IndicatorPropertiesApmAvailability:
+		i, _ := ind.(*slo.IndicatorPropertiesApmAvailability) //nolint:all
+		ret.IndicatorPropertiesApmAvailability = i
+
+	case *slo.IndicatorPropertiesApmLatency:
+		i, _ := ind.(*slo.IndicatorPropertiesApmLatency) //nolint:all
+		ret.IndicatorPropertiesApmLatency = i
+
+	case *slo.IndicatorPropertiesCustomKql:
+		i, _ := ind.(*slo.IndicatorPropertiesCustomKql) //nolint:all
+		ret.IndicatorPropertiesCustomKql = i
+
+	case *slo.IndicatorPropertiesCustomMetric:
+		i, _ := ind.(*slo.IndicatorPropertiesCustomMetric) //nolint:all
+		ret.IndicatorPropertiesCustomMetric = i
+
+	case *slo.IndicatorPropertiesHistogram:
+		i, _ := ind.(*slo.IndicatorPropertiesHistogram) //nolint:all
+		ret.IndicatorPropertiesHistogram = i
+
+	default:
+		return ret, fmt.Errorf("unknown indicator type: %T", ind)
+	}
+
+	return ret, nil
+}
+
+func sloResponseToModel(spaceID string, res *slo.SloResponse) *models.Slo {
+	if res == nil {
+		return nil
+	}
+
+	return &models.Slo{
```

Comment:
<img width="635" alt="image" src="https://github.com/elastic/terraform-provider-elasticstack/assets/597812/be4bbd23-f262-4ffe-9fa1-8cc5b1d23af0">
From the model I suppose. Most of these fields except for settings are compulsory according to the spec. Settings is a pointer so if its nil, it should be ok right?

Reply:
Non of them are marked as required in the openapi spec, otherwise they wouldn't be pointers in the generated code. I'd probably update our internal model to accept the optionality or explicitly deal with it here. Leaving it like this will create provider crashes.

Reply:
Can you elaborate here? OpenAPI spec shows everything but settings and id are required. 
<img width="827" alt="image" src="https://github.com/elastic/terraform-provider-elasticstack/assets/597812/d4b5f902-e3f9-4cbd-a8ae-f2542e448b9e">

Reply:
So this is interesting, should our internal model reflect the `Create` or `Read` version?
eg `CreateSloRequest` vs `SloResponse`

Reply:
The request, it's what a user needs to provide to create the resource.

Reply:
This is fixed by making the response have required fields

Reply:
https://github.com/elastic/kibana/pull/163430